### PR TITLE
HasStateChanged - deprecated on MSAL, quick fix on ADAL

### PR DESCRIPTION
**Pull request by bgavrilMS**
ADAL fix for https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/issues/1186

### CODE REVIEW

File: msal/src/Microsoft.Identity.Client/TokenCache.cs [hunk 1]
```diff
@@ -127,6 +126,9 @@ public TokenCache()
         /// MSAL methods set this flag after any change.
         /// Caller applications should reset the flag after serializing and persisting the state of the cache.
         /// </summary>
+        [Obsolete("Please use the equivalent flag TokenCacheNotificationArgs.HasStateChanged, " +
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs:129]: @jmprieur - please review the Obsolete message

File: msal/src/Microsoft.Identity.Client/TokenCache.cs [hunk 2]
```diff
@@ -199,999 +204,1026 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
                         Account = msalAccessTokenCacheItem.HomeAccountId != null ?
                                     new Account(msalAccessTokenCacheItem.HomeAccountId,
                                     preferredUsername, preferredEnvironmentHost) :
-                                    null
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs:202]: Apollogies for the messy diff, the only changes made here are - whenever the old flag HasStateChanged was set to true: 

- add #pragma warning disable CS0618 // Type or member is obsolete
- make the args flag true

File: msal/src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs [hunk 3]
```diff
@@ -48,5 +48,11 @@ public sealed partial class TokenCacheNotificationArgs
         /// Gets the account involved in the cache transaction.
         /// </summary>
         public IAccount Account { get; internal set; }
+
+        /// <summary>
+        /// Indicates whether the state of the cache has changed, for example when tokens are being added or removed. 
+        /// Not all cache operations modify the state of the cache.
+        /// </summary>
+        public bool HasStateChanged { get; internal set; }
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs:56]: @jmprieur - please review public API comment

File: msal/tests/AutomationApp/TokenCacheDelegates.cs [hunk 4]
```diff
@@ -1,65 +0,0 @@
...
-//
-//------------------------------------------------------------------------------
-
-using System.IO;
-using System.Security.Cryptography;
-using Microsoft.Identity.Client;
-
-namespace AutomationApp
-{
-    internal static class TokenCacheDelegates
```
Comment (bgavrilMS) [inline msal/tests/AutomationApp/TokenCacheDelegates.cs:34]: Not used

File: msal/src/Microsoft.Identity.Client/TokenCache.cs (hunk 1 above)
Comment (jmprieur) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs:129]: read-only? you mean modifies the cache?

**Review by MarkZuber — APPROVED** [review]

File: msal/tests/AutomationApp/TokenCacheDelegates.cs (hunk 4 above)
Comment (henrik-me) [reply msal/tests/AutomationApp/TokenCacheDelegates.cs:34]: should it be? (test gap?)

---
In reply to: [236806734](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1436#discussion_r236806734) [](ancestors = 236806734)

**Review by henrik-me — APPROVED** [review]
:shipit:

**Pull request merged**